server/src/apub/shared_inbox.rs

```diff
@@ -0,0 +1,241 @@
+use super::*;
+
+#[serde(untagged)]
+#[derive(Serialize, Deserialize, Debug)]
+pub enum SharedAcceptedObjects {
+  Create(Create),
+  Update(Update),
+}
+
+/// Handler for all incoming activities to user inboxes.
+pub async fn shared_inbox(
+  request: HttpRequest,
+  input: web::Json<SharedAcceptedObjects>,
+  db: DbPoolParam,
+  chat_server: ChatServerParam,
+) -> Result<HttpResponse, Error> {
+  // TODO: would be nice if we could do the signature check here, but we cant access the actor property
+  let input = input.into_inner();
+  let conn = &db.get().unwrap();
+
+  let json = serde_json::to_string(&input)?;
+  debug!("Shared inbox received activity: {:?}", &json);
+
+  match input {
+    SharedAcceptedObjects::Create(c) => handle_create(&c, &request, &conn, chat_server),
+    SharedAcceptedObjects::Update(u) => handle_update(&u, &request, &conn, chat_server),
+  }
+}
+
+/// Handle create activities and insert them in the database.
+fn handle_create(
+  create: &Create,
+  request: &HttpRequest,
+  conn: &PgConnection,
+  chat_server: ChatServerParam,
+) -> Result<HttpResponse, Error> {
+  let base_box = create.create_props.get_object_base_box().unwrap();
+
+  if base_box.is_kind(PageType) {
+    let page = create
+      .create_props
+      .get_object_base_box()
+      .to_owned()
+      .unwrap()
+      .to_owned()
+      .to_concrete::<Page>()?;
```

Comment:
@asonix Also I noticed that your activitystreams library has `is_kind`, but not `get_kind` / `kind`, so I can't do a `match` on the kind, I have to do if `is_kind`.

Reply:
that is not true. there is a `.kind()` method on base_box

Reply:
it returns an `Option<&str>`, which you can match on